### PR TITLE
Reorganize a bit the structure of the wall files

### PR DESCRIPTION
**Pull request by tulioricci**
Some cosmetic changes and initial steps towards a more thorough refactor of the gas-wall structure (#930) 
- some renaming regarding the wall calculations that I believe to make more sense now
- moved the classes that are used for heat conduction only from driver to `wall_model`
- moved the `mask_from_elements` to `utils`
- new fluid state encapsulation `PorousFlowFluidState`  to uncouple the wall density from `dv` to the new `wv`, as well as existing wall dependent vars `wdv` to `wv` 
- new GasModel-like encapsulation `PorousFlowModel` that considers the wall thermophysical properties.
- new transport class `PorousWallTransport` coupling gas and wall properties
- everything that is used exclusively by the `ablation_workshop` example is inside that driver. Now the material files are cleaner and more "prediction targeted"

**Questions for the review** @MTCam:

After quick first pass, looks good to me, just need to have a more detailed look to check the rest of the boxes.

- [x] Is the scope and purpose of the PR clear?
  - [x] The PR should have a description.
  - [x] The PR should have a guide if needed (e.g., an ordering).
- [x] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

### CODE REVIEW

File: mirgecom/wall_model.py
Comment (majosm) [inline mirgecom/wall_model.py:273]: The `Model`/`EOS` terminology here is confusing me a little (especially with one inheriting from the other). 

Questions:

1. Is the base class needed? i.e., will there be more than one concrete class along the lines of `PorousFlowModel`?
2. If yes, should `PorousFlowModel` have a more specific name? It sounds kind of generic at the moment.

**Review by majosm — COMMENTED** [review]
Added some thoughts. It was a little tricky to review this due to the intermediate-ness of some of the changes, heh. I also still don't have a great mental image of the new classes and how they all relate to each other; seeing a diagram of the hierarchy might be helpful.

File: doc/operators/thermally_coupled.rst
Comment (tulioricci) [reply doc/operators/thermally_coupled.rst:51]: IIUC about splitting the operator docs, I completely agree but I would leave that to a new PR.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py:357]: It can be the `PorousFlowModel`. This is one of the main differences from the previous way where wall and gas were mixed together.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py:445]: Done.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py:11]: Done.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py:150]: Done.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py:80]: Done.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py:410]: Hmm what? The FIXME? I don't see a way to pass `niter` to `get_temperature` since it is inside `make_fluid_state`.

File: mirgecom/materials/initializer.py
Comment (tulioricci) [reply mirgecom/materials/initializer.py:111]: I was following the fluid `initializers` but I don't see why not if you prefer functions.

File: mirgecom/simutil.py
Comment (tulioricci) [reply mirgecom/simutil.py:93]: Done.

File: mirgecom/simutil.py
Comment (tulioricci) [reply mirgecom/simutil.py:107]: Done.

File: mirgecom/transport.py
Comment (tulioricci) [reply mirgecom/transport.py:708]: Done.

File: mirgecom/gas_model.py
Comment (majosm) [reply mirgecom/gas_model.py:357]: The point I was trying to make is that in principle if it's called `gas_model` and documented as a `GasModel`, it should be something that derives from `GasModel`. I'm OK with weakening that if it's an intermediate step towards #930, but I would add a FIXME here explaining what's going on in order to make it less alarming.

File: mirgecom/gas_model.py
Comment (majosm) [reply mirgecom/gas_model.py:410]: Would it make sense to specify that when you create the gas model? Or do you want more fine-grained control?

File: mirgecom/materials/initializer.py
Comment (majosm) [reply mirgecom/materials/initializer.py:111]: 🤷‍♂️ I don't remember exactly why we did that with the fluid initializers. (Maybe it was so we could group them together with the `exact_*` functions? Though we could probably accomplish the same thing using pymbolic instead nowadays.) I guess it's fine to leave them as classes; we can always change it later if we want.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py:357]: It is an intermediate step. I would rename everything from `gas_model` to something like `flow_model`. Since everything in the code calls `make_fluid_state`, I am avoiding any major changes/refactor now.

I will add a FIXME/TODO to clarify this point and as a future reminder.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py:432]: I (tried to) follow the same structure of the fluid-only: EOS handles the physical definition of everything, and the model is (kinda of) a wrapper/encompassing class. Additional functions are required since we have to combine both solid and gas (for instance, the temperature). The `model` allows us to access everything, while its sub-classes are related to the different aspects (gas, wall, transport).

Since they are all wrapped together, it should be possible to access the mother/sibling classes without explicitly giving it as an argument, but I couldn't make it work.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py:436]: Done.

File: mirgecom/materials/initializer.py
Comment (tulioricci) [reply mirgecom/materials/initializer.py:111]: I will leave as classes then.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py:292]: My main point for a major refactor is that we use `gas_model` everywhere and can't easily decouple totally fluid-only and solid+fluid because of that. Here, I follow the increasing complexity of inviscid -> viscous -> porous flow and I am happy with it.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py:252]: Can't inherit because pylint was complaining since I don't assign all functions. However, renamed to GasEOS instead of GasProperties.

File: mirgecom/wall_model.py
Comment (tulioricci) [reply mirgecom/wall_model.py:50]: I will keep solid for everything non-porous/conductive only and double-check that everything else doesn't have it.

File: mirgecom/materials/initializer.py
Comment (tulioricci) [reply mirgecom/materials/initializer.py:36]: I will keep solid for everything non-porous/conductive only and double-check that everything else doesn't have it. Ok?

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py:225]: Done.

File: mirgecom/wall_model.py
Comment (tulioricci) [reply mirgecom/wall_model.py:273]: I used `EOS` as the abstract class and the `model` as the wrapper class that combines both solid and gas (thus, a different EOS for the interacting solid+gas). The feeding classes (eos, transport and wall_model) are specific to gas, transport model and wall-only properties, respectively.

To clarify the individual classes, I added more details on the docs. If you have suggestions for variables/classes name in order to clarify, please!

1. I don't know if there will be, but it may be possible.
2. I used this name to keep the `model` in the nomenclature as in `GasModel`.

Let me know if it is still confusing.

File: doc/operators/thermally_coupled.rst
Comment (tulioricci) [reply doc/operators/thermally_coupled.rst:39]: Done. And added a new tab with wall model only in the docs.

File: doc/operators/thermally_coupled.rst
Comment (tulioricci) [reply doc/operators/thermally_coupled.rst:51]: Added a new tab with wall model only. I like the idea of reducing the operators. I can do that over the next days.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py]: IIUC, done.

File: mirgecom/gas_model.py
Comment (majosm) [reply mirgecom/gas_model.py:443]: After taking a closer look: you're just passing the `gas_model` in to get access to the EOS and wall models, right? Could you pass those into the transport model's `__init__` and store them instead?

File: mirgecom/multiphysics/__init__.py
Comment (majosm) [inline mirgecom/multiphysics/__init__.py:31]: (Unrelated to your change, but while scrolling through I noticed the stray merge conflict annotation here. Could you delete it? 🙂)

File: mirgecom/transport.py
Comment (majosm) [inline mirgecom/transport.py:703]: I think this can't (or at least shouldn't) inherit from `TransportModel`, because it doesn't implement the same interface (the methods take different arguments). I suspect we could generalize the transport model interface later on to make it compatible. Maybe for now add a `# FIXME: Generalize TransportModel interface to accept state variables other than fluid cv`.

File: mirgecom/wall_model.py
Comment (majosm) [reply mirgecom/wall_model.py:319]: I would maybe drop the base class for now (unless there's a current use for it that I'm not quite understanding). Calling it an `EOS` is kind of confusing in a mirgecom context since the EOS is usually a member object of the `Model`, not a base class of it (or both, in this case).

**Comment by majosm** [comment]
I forgot to mention: I have an idea to simplify `make_fluid_state` a little. I'll give it a shot, and if it pans out I'll make a PR (I won't let that hold up this PR though).

File: mirgecom/transport.py
Comment (tulioricci) [reply mirgecom/transport.py]: Done. No more inheritance and FIXME added.

File: mirgecom/multiphysics/__init__.py
Comment (tulioricci) [reply mirgecom/multiphysics/__init__.py]: Done.

File: mirgecom/gas_model.py
Comment (tulioricci) [reply mirgecom/gas_model.py]: Done.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py]: Modified the example to follow the same structure from `gas_model`. Now it has its own transport class.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py]: Added a local transport class to overcome this. Related to your next comment below as well.

File: examples/ablation-workshop-mpi.py
Comment (tulioricci) [reply examples/ablation-workshop-mpi.py]: Implemented the full EOS.

**Comment by tulioricci** [comment]
Hi @majosm, I think I covered all the points you raised. I couldn't reply directly wrt the `eos` and `wall_eos` in the transport variables, but with the latest changes, now it has the same structure of pure gas. I don't know if it is worth to remove one argument but add the classes in the init, as you suggested. It doesn't look like a significant improvement.

Also, note that I renamed `wall_model` in `PorousFlowModel` to `wall_eos` to keep consistency with the gas nomenclature. Hopefully this also helps makes things more clear.

File: mirgecom/eos.py
Comment (MTCam) [inline mirgecom/eos.py:667]:
```suggestion
    def gas_const(self, cv: ConservedVars,  # type: ignore[override]
                  temperature: Optional[DOFArray] = None):
```

File: mirgecom/materials/carbon_fiber.py
Comment (MTCam) [inline mirgecom/materials/carbon_fiber.py:52]: tbh, I do not see any real harm in internalizing some of our use-specific constructs.  If it makes it easier for us (and others) to use, then we should.  Is that the question, or did I miss the mark?

**Review by MTCam — COMMENTED** [review]
👍  Looks like a net positive change to me.  It is hard to tell what all the implications are, and the unintended effects.  We'll just have to _use_ it and find out, heh.

File: mirgecom/gas_model.py
Comment (MTCam) [reply mirgecom/gas_model.py]: If niter changes, there will be re-compiles.   Just FYI.

File: mirgecom/materials/carbon_fiber.py
Comment (tulioricci) [reply mirgecom/materials/carbon_fiber.py:52]: The original comment was originally here: https://github.com/illinois-ceesd/mirgecom/pull/875#discussion_r1261414281

Just keeping in the code as a reminder for the coupling PR which will effectively use the models. I suggest we postpone the discussion on this.

File: examples/ablation-workshop.py
Comment (tulioricci) [reply examples/ablation-workshop.py:250]: The arguments for the functions are different, that is why we removed the inheritance and added the FIXME.

File: examples/ablation-workshop.py
Comment (tulioricci) [reply examples/ablation-workshop.py:346]: Since we are inheriting the class, I removed all the automethods because they seem superfluous.

File: mirgecom/wall_model.py
Comment (tulioricci) [reply mirgecom/wall_model.py:304]: Same as above

File: mirgecom/wall_model.py
Comment (tulioricci) [reply mirgecom/wall_model.py:424]: Changed. But in my mind it is always going to recompile if changing the number of iterations because of the DAG construction, or am I wrong?

Comment (majosm) [reply mirgecom/wall_model.py]: Yep, that's right. Putting it in the constructor maybe just makes it slightly less tempting to try to "optimize" the number of iterations by using different values in different calls.

**Review by majosm — APPROVED** [review]
👍 Looks good on my end.

**Pull request merged**
